rules/falco_rules.yaml

```diff
@@ -3333,6 +3333,17 @@
   priority: WARNING
   tags: [host, container, mitre_execution, mitre_defense_evasion, T1036.005, T1059.004]
 
+- rule: Drop and execute new binary in container
+  desc:
+    Detect if an executable not belonging to the base image of a container is being executed.
+    The drop and execute pattern can be observed very often after an attacker gained an initial foothold.
+    is_exe_upper_layer filter field only applies for container runtimes that use overlayfs as union mount filesystem.
+  condition: >
+    spawned_process and container and proc.is_exe_upper_layer=true
```

Comment:
In terms of filtering perhaps check if it would make sense to also filter by `%proc.exe_ino.ctime_duration_pidns_start` (it is the time delta in nanoseconds between container start and ctime of the exe_file)? For example, exclude the first n nanoseconds of the lifetime of a container as containers may pull artifacts or do bunch of symlinking right after start up (symlinks also create a new inode aka a new executable in our definition / rule here).

Reply:
This is something I considered. I would say let's start with not cutting the beginning of container life. I still believe in most use cases containers will be immutable. And in any case, how to choose the threshold `n`? What we can do instead, is allow known containers to pull and execute artifacts within n nanoseconds by means of a whitelist that we can introduce later to fine-tune the rule if it becomes too noisy :)

Moreover on the symlink point: `exe_upper_layer` as you know retrieves the inode of the executable from the `mm_struct->exe_file` from the process descriptor (task_struct). This is something that the kernel already resolved for us, so what we are taking here in case of symlink is the real inode the symlink is pointing to. So a symlink to something that is part of the base image will not cause the rule to trigger! I need to double check this, but I am pretty sure it works like this

Reply:
re tuning agreed, also think for the most part containers should be immutable and when they are not the tuning may be too organization and application specific hence similar to other rules, the end user can customize the rule. And what you suggested with "known containers to pull and execute artifacts within n nanoseconds" sounds great. Maybe instead of adding tuning already just list a few options as comment above? Consider this feedback [nit], no need to address. Finding the right thresholds and such yeah that is the tricky part. Also I forgot to implement something like `ctime - mtime` ... a parameter I noticed can contain high value info as well.

re symlinks at runtime let me double check as well.

Reply:
> Moreover on the symlink point: `exe_upper_layer` as you know retrieves the inode of the executable from the `mm_struct->exe_file` from the process descriptor (task_struct). This is something that the kernel already resolved for us, so what we are taking here in case of symlink is the real inode the symlink is pointing to. So a symlink to something that is part of the base image will not cause the rule to trigger! I need to double check this, but I am pretty sure it works like this


correct, also just verified that if you dummy symlink 2 binaries that were already in the container -> not in upper layer and what you describe is how you can detect for example ` /bin/uname -> /bin/new_malicious`, but your exepath would be `/bin/uname`, see https://github.com/falcosecurity/libs/pull/705#issuecomment-1312627375 I think that's what would be intended to be detected in such cases right?

Reply:
Strongly agreed with `ctime - mtime`, this could be particularly relevant!
Not sure what you mean by the second point, but what I was saying is that no matter how you deal with symlinks, if you symlink to something that was in the base image the rule doesn't have to trigger (and it does not), else it has to trigger (this case is correctly handled). Of course, using symlinks the exepath could potentially be wrong due to the fact that we retrieve it from userspace and use it unresolved (at least for now!) 🙂 Did I get your point right or not?

Reply:
@loresuso re your summary yup that's about it what I meant. It may seem obvious to us, but for some folks reading this it may improve the clarity of what this detection captures.

Re `ctime - mtime` more longer term I just opened this https://github.com/falcosecurity/falco/issues/2403 so maybe in the future we will be more flexibility for such cases without needed to introduce a new filtercheck field in the source.


---

Back to the detection: I think the tuning really will be very application and organization specific - just like many of the other rules. The biggest tuning win likely is profiled containers/apps combined w/ process lineage information as we discussed before, so maybe this as template lists or hint could make sense, and then it can be further tuned after field experience. It definitely won't explode the data pipeline for sure would say.

Reply:
Nice! 
Added a list and commented on it so that users can already use it if they need. Thanks!